lua/wire/stools/expression2.lua

```diff
@@ -55,20 +72,135 @@ if SERVER then
+
+	-- Simple serverside only local table for storing view requests to make handling them not spaghetti code
+	local viewRequests = {}
+
+	hook.Add("Tick", "WireExpression2_InvalidateRequests", function()
+		for k, v in pairs(viewRequests) do
+			if not IsValid(v.initiator) or not IsValid(v.owner) or not IsValid(v.chip) then
+				-- Void the request if any of the entities are no longer valid
+				viewRequests[k] = nil
+			elseif CurTime() > v.expiry then
+				BetterChatPrint(v.initiator, "Your request to view "..v.owner:Nick().."'s chip, '"..v.name.."', has expired")
+				viewRequests[k] = nil
+			end
+		end
+	end)
 
-		if IsValid(trace.Entity) and trace.Entity:GetClass() == "gmod_wire_expression2" then
-			self:Download(player, trace.Entity)
-			return true
+	local function RequestView(chip, initiator)
+		local index = chip:EntIndex()
+		local truncName = string.sub(chip.name, 1, 256) -- In case someone starts making cursed names
+
+		-- Make sure this isn't creating a request for a chip with an outstanding valid request
+		if viewRequests[index] and viewRequests[index].expiry > CurTime() then
+			BetterChatPrint(initiator, "Request to view '"..truncName.."' already sent")
+			return
 		end
 
-		net.Start("WireExpression2_OpenEditor") net.Send(player)
-		return false
+		-- Otherwise, print to the tool user's chat that a view request was sent
+		-- and send a view request to the chip's owner (also print a message to their chat to tell them they received a request)
+		BetterChatPrint(initiator, "E2 view request sent for '"..truncName.."' owned by "..chip.player:Nick())
+		BetterChatPrint(chip.player, "You just received a request to view your E2 '"..truncName.."' from "..initiator:Nick()..", which you can view in your context menu")
+
+		-- Add the request data to the local requests table for when the request from the client comes in
+		viewRequests[index] = {
+			initiator = initiator,
+			owner = chip.player,
+			name = truncName,
+			chip = chip,
+			expiry = CurTime() + 60 -- 1 minute for the request before it's invalidated (could make this a convar)
+		}
+
+		net.Start("WireExpression2_ViewRequest")
+			net.WriteInt(chip:EntIndex(), 32) -- For lookup in the request table when the owner answers
+			net.WriteFloat(CurTime() + 60)            -- For making requests expire on time clientside
+			net.WriteEntity(initiator)                -- The player attempting to view the E2
+			net.WriteString(truncName)                -- Name of the E2 so the owner knows what they're agreeing to
+			net.WriteEntity(chip)             -- Chip entity for validation clientside
+		net.Send(chip.player)
+	end
+
+	util.AddNetworkString("WireExpression2_OpenEditor")
+	util.AddNetworkString("WireExpression2_ViewRequest")
+	util.AddNetworkString("WireExpression2_AnswerRequest")
+	function TOOL:Think()
+		--[[
+			I had to replace TOOL:RightClick with TOOL:Think as prop protection was preventing
+			the view requests system from functioning as intended
```

Comment:
This should have been a big hint to you early on that you're doing it wrong...

Reply:
Well if it requires you to give full prop protection perms on something to use view requests, kinda defeats the point.

Reply:
Would you rather him overload the gmod_tool functions?

Reply:
The view requests almost completely replace the flow for both tool and remote updater code downloading, so I’m not sure what you mean by a tool *option* or the requirement for an additional tool.